Create blert v0.1.0

frolv opened the pull request:
Blert is a PvM tracking and data analysis system hosted at https://blert.io.

This plugin collects data about players' raids and transmits it to the Blert servers to be processed and visualized.

frolv commented:
This is a pretty large and complex project, so I'd be happy to discuss it in more detail as it is reviewed.

frolv commented:
This should be ready to review now. Happy to chat about anything, and please reach out before merging.

raiyni commented:
The chances this gets reviewed in a timely manor is very low based on the number of dependencies added

frolv commented:
No worries, I don't expect a quick review. Was just pointing out that I don't intend to make any further commits.

Nightfirecat commented:
[This would make the build non-reproduceable](https://github.com/blert-io/plugin/blob/e58bf32ecd9b5f990d469cb0026572e06f9516f2/build.gradle#L60), please don't include it.

frolv commented:
> [This would make the build non-reproduceable](https://github.com/blert-io/plugin/blob/e58bf32ecd9b5f990d469cb0026572e06f9516f2/build.gradle#L60), please don't include it.

Removed the remote check, it wasn't actually being used.

frolv commented on package/verification-template/build.gradle:
```diff
@@ -117,6 +117,24 @@ dependencies {
+	thirdParty("com.google.gradle:osdetector-gradle-plugin:1.7.3") {
+		because "blert"
+	}
+	thirdParty("kr.motd.maven:os-maven-plugin:1.7.1") {
```
I tried looking into it but couldn't find exactly where this is pulled in -- I assume it's a transitive of one of the protobuf libraries.

frolv replied:
Oh, I was looking the in wrong place. It is indeed coming from protobuf:
```
+--- com.google.protobuf:protobuf-gradle-plugin:0.9.4
|    \--- com.google.gradle:osdetector-gradle-plugin:1.7.3
|         \--- kr.motd.maven:os-maven-plugin:1.7.1
|              \--- com.google.code.findbugs:jsr305:3.0.2
```

frolv commented:
On the dependency issue:

The plugin pulls in a single direct dependency (the protobuf compiler), which is quite large and has a bunch of transitives. However, this dependency is only required at build time to generate Java code from the plugin's `.proto` files. At runtime, the plugin reuses Runelite's existing version of the protobuf libraries and has no additional dependencies.

It is possible to remove the protoc dependency by manually running the compiler and checking in the generated Java protobuf code. This currently amounts to ~14.5K LOC (26K lines total) for serializing/deserializing protobuf messages to Java objects.

Does Runelite have a policy around checking in/reviewing generated code of this nature, and would it be preferable to eliminate the protoc dependency by doing so?

frolv commented:
I rewrote the websocket protocol to use JSON as the wire format, eliminating the protobuf dependency.

riktenx commented:
At a glance:

* Don't specify your own gson dependency at all, it should be available transitively through Runelite
* We don't allow reflection in hub plugins.

frolv commented:
Updated to not use `reflect.Type` and remove gson dep. Latest commit doesn't have checked in proto files.

frolv commented:
Awesome, thank you!

riktenx commented:
blocking
* update your warning message per the above
* cancel() WebSocketEventHandler.requestTimeout on plugin shutdown

nonblocking
* i'm guessing no but can you still get events on the WebSocketListener after you call close?
* won't closeWebsocketClient deadlock since you're get()ing the close future from a message callback that I assume is invoked in the websocket's own thread?
* we really do not like conditional eventbus registering because it makes it very easy to leak registrations and have your plugin continue to act in unexpected ways (potentially even after a player disables the plugin). I've traced the way you to do this to the best of my ability and it seems stable.

frolv commented:
> update your warning message per the above

updated

> cancel() WebSocketEventHandler.requestTimeout on plugin shutdown

Added through WebSocketManager's close

> i'm guessing no but can you still get events on the WebSocketListener after you call close?

Shouldn't be possible but I added an extra guard to not process events in CLOSING either

> won't closeWebsocketClient deadlock since you're get()ing the close future from a message callback that I assume is invoked in the websocket's own thread?

Good catch, switched to nonblocking close

> we really do not like conditional eventbus registering because it makes it very easy to leak registrations

I can take a look at rearchitecting how challenge/data tracker event handling is managed in a followup update.
